src/Model/HasAttributesTrait.php

```diff
@@ -0,0 +1,238 @@
+                $object = new $class();
+
+                foreach ($originalAttributes as $key => $value) {
+                    $object->setDocumentAttribute($key, $value);
+                }
+            }
+        }
+
+        foreach ($input as $key => $value) {
+            if ($force
+                || ((!$object->fillable || in_array($key, $object->fillable)) && !in_array($key, $object->guarded))) {
+                if ($value instanceof stdClass) {
+                    $value = json_decode(json_encode($value), true); // cast to array
+                }
+
+                $object->setDocumentAttribute($key, $value);
+            }
+        }
+
+        return $object;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasDocumentAttribute(string $key): bool
+    {
+        return !is_null($this->getDocumentAttribute($key));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function &getDocumentAttribute(string $key)
+    {
+        if ($this->mutable && $this->hasMutatorMethod($key, 'get')) {
+            $this->mutableCache[$key] = $this->{$this->buildMutatorMethod($key, 'get')}();
+
+            return $this->mutableCache[$key];
+        }
+
+        if (array_key_exists($key, $this->attributes)) {
+            return $this->attributes[$key];
+        }
+
+        if (!method_exists(self::class, $key) && method_exists($this, $key)) {
+            return $this->getRelationResults($key);
+        }
+
+        $this->attributes[$key] = null;
+
+        return $this->attributes[$key];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDocumentAttributes(): array
+    {
+        foreach ($this->attributes as $field => $value) {
+            if (null === $value) {
+                $this->cleanDocumentAttribute($field);
+            }
+        }
+
+        return $this->attributes ?? [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function cleanDocumentAttribute(string $key)
+    {
+        unset($this->attributes[$key]);
+
+        if ($this->hasFieldRelation($key)) {
+            $this->unsetRelation($this->getFieldRelation($key));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setDocumentAttribute(string $key, $value)
+    {
+        if ($this->mutable && $this->hasMutatorMethod($key, 'set')) {
+            $value = $this->{$this->buildMutatorMethod($key, 'set')}($value);
+        }
+
+        if (null === $value) {
+            $this->cleanDocumentAttribute($key);
+
+            return;
+        }
+
+        $this->attributes[$key] = $value;
+
+        if ($this->hasFieldRelation($key)) {
+            $this->unsetRelation($this->getFieldRelation($key));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function syncOriginalDocumentAttributes()
+    {
+        try {
+            $this->originalAttributes = unserialize(serialize($this->getDocumentAttributes()));
```

Comment:
note to self: this takes way too many memory, think about alternatives or drop it altogether.